notebooks/03_evaluate/comparison_movielens.ipynb

```diff
@@ -0,0 +1,982 @@
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_summary(data, algo, k, train_time, time_rating=np.nan, rating_metrics=None, time_ranking=np.nan, ranking_metrics=None):\n",
+    "    summary = {\"Data\": data, \"Algo\": algo, \"K\": k, \"Train time\": train_time, \"Rating time\": time_rating, \"Ranking time\": time_ranking}\n",
+    "    if rating_metrics is None:\n",
+    "        rating_metrics = {\n",
+    "            \"RMSE\": np.nan,\n",
+    "            \"MAE\": np.nan,\n",
+    "            \"R2\": np.nan,\n",
+    "            \"Explained Variance\": np.nan,\n",
+    "        }\n",
+    "    if ranking_metrics is None:\n",
+    "        ranking_metrics = {\n",
+    "            \"MAP\": np.nan,\n",
+    "            \"nDCG@k\": np.nan,\n",
+    "            \"Precision@k\": np.nan,\n",
+    "            \"Recall@k\": np.nan,\n",
+    "        }\n",
+    "    summary.update(rating_metrics)\n",
+    "    summary.update(ranking_metrics)\n",
+    "    return summary"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data_sizes = [\"100k\"] # Movielens data size: 100k, 1m, 10m, or 20m\n",
+    "algorithms = [\"als\", \"svd\", \"sar_single_node\", \"fastai\", \"ncf\"]\n",
+    "algorithms = [\"fastai\"]\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Size of Movielens 100k: (100000, 4)\n",
+      "Train set size: (74992, 4)\n",
+      "Test set size: (25008, 4)\n",
+      "\n",
+      "Computing fastai algorithm on Movielens 100k\n",
+      "Training time: 0:00:20.174065\n",
+      "\n",
+      "Evaluating with fastai\n",
+      "Rating prediction time: 0:00:00.061902\n",
+      "Rating metrics: \n",
+      "{\n",
+      "    \"Explained Variance\": 0.36284020679183915,\n",
+      "    \"MAE\": 0.7713322553785931,\n",
+      "    \"R2\": 0.3626695202664241,\n",
+      "    \"RMSE\": 0.9629789752608614\n",
+      "}\n",
+      "Ranking prediction time: 0:00:02.700339\n",
+      "Ranking metrics: \n",
+      "{\n",
+      "    \"MAP\": 0.009747592680945431,\n",
+      "    \"Precision@k\": 0.04729586426299045,\n",
+      "    \"Recall@k\": 0.02708837308673072,\n",
+      "    \"nDCG@k\": 0.0499251200590822\n",
+      "}\n",
+      "CPU times: user 32.2 s, sys: 4.82 s, total: 37 s\n",
+      "Wall time: 40.3 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "\n",
+    "# For each data size and each algorithm, a recommender is evaluated. \n",
+    "cols = [\"Data\", \"Algo\", \"K\", \"Train time\", \"Rating time\", \"RMSE\", \"MAE\", \"R2\", \"Explained Variance\", \"Ranking time\", \"MAP\", \"nDCG@k\", \"Precision@k\", \"Recall@k\", ]\n",
```

Comment:
Here the cols should be the same as the dict in `generate_summary()`, so better to use a common variable.